embedded/sql/engine_test.go

```diff
@@ -341,6 +343,198 @@ func TestTimestampCasts(t *testing.T) {
+			"INSERT INTO float_table(ft) VALUES(@ft)", map[string]interface{}{
+				"ft": 0.78,
+			},
+			nil,
+		)
+		require.NoError(t, err)
+
+		r, err := engine.Query("SELECT ft FROM float_table WHERE ft = @ft ORDER BY id", map[string]interface{}{
+			"ft": 0.78,
+		}, nil)
+		require.NoError(t, err)
+
+		row, err := r.Read()
+		require.NoError(t, err)
+		require.Equal(t, FloatType, row.Values[sel].Type())
+		require.Equal(t, 0.78, row.Values[sel].Value())
+
+		_, err = r.Read()
+		require.ErrorIs(t, err, ErrNoMoreRows)
+
+		err = r.Close()
+		require.NoError(t, err)
+
+		r, err = engine.Query("SELECT ts FROM float_table WHERE ft = @ft ORDER BY id", map[string]interface{}{
+			"ft": "2021-12-06 10:14",
+		}, nil)
+		require.NoError(t, err)
+
+		_, err = r.Read()
+		require.ErrorIs(t, err, ErrNotComparableValues)
+
+		err = r.Close()
+		require.NoError(t, err)
+	})
+}
+
+func TestFloatIndex(t *testing.T) {
+	st, err := store.Open("float_index", store.DefaultOptions())
+	require.NoError(t, err)
+	defer st.Close()
+	defer os.RemoveAll("float_index")
+
+	engine, err := NewEngine(st, DefaultOptions().WithPrefix(sqlPrefix))
+	require.NoError(t, err)
+
+	_, _, err = engine.Exec("CREATE DATABASE db1", nil, nil)
+	require.NoError(t, err)
+
+	err = engine.SetDefaultDatabase("db1")
+	require.NoError(t, err)
+
+	_, _, err = engine.Exec("CREATE TABLE IF NOT EXISTS float_index (id INTEGER AUTO_INCREMENT, ft FLOAT, PRIMARY KEY id)", nil, nil)
+	require.NoError(t, err)
+
+	_, _, err = engine.Exec("CREATE INDEX ON float_index(ft)", nil, nil)
+	require.NoError(t, err)
+
+	for i := 100; i > 0; i-- {
+		val, _ := strconv.ParseFloat(fmt.Sprint(i, ".", i), 64)
+		_, _, err = engine.Exec("INSERT INTO float_index(ft) VALUES(@ft)", map[string]interface{}{"ft": val}, nil)
+		require.NoError(t, err)
+	}
+
+	r, err := engine.Query("SELECT * FROM float_index ORDER BY ft", nil, nil)
+	require.NoError(t, err)
+	defer r.Close()
+
+	for i := 100; i > 0; i-- {
+		row, err := r.Read()
+		require.NoError(t, err)
+		require.EqualValues(t, i, row.Values[EncodeSelector("", "db1", "float_index", "id")].Value())
+	}
+
+	_, err = r.Read()
+	require.ErrorIs(t, err, ErrNoMoreRows)
+}
+
+func TestFloatIndexOnNegatives(t *testing.T) {
+	st, err := store.Open("float_index", store.DefaultOptions())
+	require.NoError(t, err)
+	defer st.Close()
+	defer os.RemoveAll("float_index")
+
+	engine, err := NewEngine(st, DefaultOptions().WithPrefix(sqlPrefix))
+	require.NoError(t, err)
+
+	_, _, err = engine.Exec("CREATE DATABASE db1", nil, nil)
+	require.NoError(t, err)
+
+	err = engine.SetDefaultDatabase("db1")
+	require.NoError(t, err)
+
+	_, _, err = engine.Exec("CREATE TABLE IF NOT EXISTS float_index (id INTEGER AUTO_INCREMENT, ft FLOAT, PRIMARY KEY id)", nil, nil)
+	require.NoError(t, err)
+
+	_, _, err = engine.Exec("CREATE INDEX ON float_index(ft)", nil, nil)
+	require.NoError(t, err)
+
+	floatSerie := []float64{-1.0, 3.345, -0.5, 0.0, -100.8, 0.5, 1.0, math.MaxFloat64, -math.MaxFloat64}
```

Comment:
Other interesting test cases to check:

* +0.0 vs -0.0 - those can have different representations in floating points although those should end up next to each other in topological sorting order (also to confirm - compiler may convert -0.0 value to 0.0 thus it should be confirmed that that's actually a negative 0 value)
* denormalized values - I think `math.SmallestNonzeroFloat64` should be such value - it should sort correctly compared to 0.0 and other numbers, also a good test would be with greatest positive denormalized value and smallest positive normalized value
* +/- Inf
* +/- NaN